src/Storages/MergeTree/ReplicatedMergeTreeQueue.cpp

```diff
@@ -169,6 +169,16 @@ void ReplicatedMergeTreeQueue::insertUnlocked(
             addPartToMutations(virtual_part_name);
     }
 
+    /// Add all source parts for mutations, to avoid replacing destination part with an empty.
+    /// (see ReplicatedMergeTreePartCheckThread::searchForMissingPartAndFetchIfPossible())
+    if (entry->type == LogEntry::MUTATE_PART)
+    {
+        for (const String & virtual_part_name : entry->source_parts)
+        {
+            virtual_parts.add(virtual_part_name, nullptr);
```

Comment:
It [has no effect](https://github.com/ClickHouse/ClickHouse/blob/1a4ccab8e6b851bdc0f4eaa84d2e4d852d04259a/src/Storages/MergeTree/ActiveDataPartSet.cpp#L29-L30), because source part is covered be `entry->new_part_name` and `entry->new_part_name` is [already added](https://github.com/ClickHouse/ClickHouse/blob/09ff66da0e026b5b1f6352e438fb0b7b7ae7d1dd/src/Storages/MergeTree/ReplicatedMergeTreeQueue.cpp#L165) to `virtual_parts`

Reply:
But `entry->new_part_name` is the destination (i.e. new part), not the source part, no?

Reply:
Oh, you mean that it is "covered"

Reply:
Right, the problem was a little bit different, fixed the patch.